src/combinatorics/additive/behrend.lean

```diff
@@ -0,0 +1,192 @@
+length `3`.
+
+The idea is that the sphere (in the `n` dimensional Euclidean space) doesn't contain arithmetic
+progressions (literally) because the corresponding ball is strictly convex. Thus we can take
+integer points on that sphere and map them onto `ℕ` in a way that preserves arithmetic progressions
+(`behrend.map`).
+
+## Main declarations
+
+* `behrend.sphere`: The intersection of the Euclidean with the positive integer quadrant. This is
+  the set that we will map on `ℕ`.
+* `behrend.map`: The map from the Euclidean space to `n`. It reads off the coordinates as digits in
+  base `d`.
+* `behrend.card_sphere_le_roth_number_nat`: Implicit lower bound on Roth numbers in terms of
+  `behrend.sphere`.
+
+## References
+
+* [Bryan Gillespie, *Behrend’s Construction*]
+  (http://www.epsilonsmall.com/resources/behrends-construction/behrend.pdf)
+* Behrend, F. A., "On sets of integers which contain no three terms in arithmetical progression"
+* [Wikipedia, *Salem-Spencer set*](https://en.wikipedia.org/wiki/Salem–Spencer_set)
+
+## Tags
+
+Salem-Spencer, Behrend construction, arithmetic progression, sphere, strictly convex
+-/
+
+open finset nat real
+open_locale big_operators pointwise
+
+namespace behrend
+variables {α β : Type*} {n d k N : ℕ} {x : fin n → ℕ}
+
+/-!
+### Turning the sphere into a Salem-Spencer set
+
+We define `behrend.sphere`, the intersection of the $$L^2$$ sphere with the positive quadrant of
+integer points. Because the $$L^2$$ closed ball is strictly convex, the $$L^2$$ sphere and
+`behrend.sphere` are Salem-Spencer (`add_salem_spencer_sphere`). Then we can turn this set in
+`fin n → ℕ` into a set in `ℕ` using `behrend.map`, which preserves `add_salem_spencer` because it is
+an additive monoid homomorphism.
+-/
+
+/-- The box `{0, ..., d - 1}^n` as a finset. -/
+def box (n d : ℕ) : finset (fin n → ℕ) := fintype.pi_finset $ λ _, range d
+
+lemma mem_box : x ∈ box n d ↔ ∀ i, x i < d := by simp only [box, fintype.mem_pi_finset, mem_range]
+
+@[simp] lemma card_box : (box n d).card = d ^ n := by simp [box]
+
+/-- The intersection of the sphere of radius `sqrt k` with the integer points in the positive
+quadrant. -/
+def sphere (n d k : ℕ) : finset (fin n → ℕ) := (box n d).filter $ λ x, ∑ i, x i^2 = k
+
+lemma sphere_zero : sphere n d 0 ⊆ 0 :=
+λ x, by simp [sphere, function.funext_iff] {contextual := tt}
+
+lemma sphere_subset_box : sphere n d k ⊆ box n d := filter_subset _ _
+
+lemma norm_of_mem_sphere {n d k : ℕ} {x : fin n → ℕ} (hx : x ∈ sphere n d k) :
+  ∥@id (euclidean_space ℝ (fin n)) (coe ∘ x : fin n → ℝ)∥ = sqrt k :=
+begin
+  rw euclidean_space.norm_eq,
+  congr',
+  change ∑ (i : fin n), ∥(x i : ℝ)∥ ^ 2 = k,
+  simp_rw [norm_coe_nat, ←cast_pow, ←cast_sum, (mem_filter.1 hx).2],
+end
+
+/-- The map that appears in Behrend's bound on Roth numbers. -/
+@[simps] def map (d : ℕ) : (fin n → ℕ) →+ ℕ :=
+{ to_fun := λ a, ∑ i, a i * d ^ (i : ℕ),
```

Comment:
This is just the "obvious" embedding of `fin n → fin d` into `fin (d ^ n)`, right? I think it would nice to have that near the equivalence we already have for `mul`.

Reply:
Done in #14817. I don't think I can use it here, however, because it would require `fin n → ℕ`, not `fin n → fin d`, and I can't switch to  `fin n → fin d` because it does not give an `add_monoid_hom` (as it loops around).

Reply:
Don't you always invoke it on elements less than `d` anyway, meaning you never hit the looping region?

Reply:
`(d - 1) + (d - 1)` loops around, though. And mostly this means that it is not an `add_monoid_hom`. Instead, it will onle be a `freiman_hom` and this is harder to prove.